--!strict
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local ItemIdEnum = require(ReplicatedStorage.Source.Enums.ItemIdEnum)
local ItemInfoType = require(ReplicatedStorage.Source.Types.ItemInfoType)
local WallStrengthConfig = require(ReplicatedStorage.Source.Configs.WallStrengthConfig)

local Items: { [ItemIdEnum.EnumType]: ItemInfoType.ItemInfoType } = {
	ScaffoldingSmall = {
		ItemId = "ScaffoldingSmall",
		DisplayName = "Small Scaffolding",
		Description = "<b>4x4 Platform</b>\nAllows temporary vertical traversal in mined-out areas.",
		Cost = 50_000,
		RebirthRequirement = 5,
		PurchaseQuantity = 1,
		MaxStack = 10,
		UsageType = "Consumable",
		ExpirationTime = 60,
		ItemCategory = "Traversal",
		DeploymentType = "GridPlacement",
		AssetPath = "Models/Items/ScaffoldingSmall",
	},
	ScaffoldingLarge = {
		ItemId = "ScaffoldingLarge",
		DisplayName = "Large Scaffolding",
		Description = "<b>8x4 Platform</b>\nAllows temporary vertical traversal in mined-out areas.",
		Cost = 100_000,
		RebirthRequirement = 8,
		PurchaseQuantity = 1,
		MaxStack = 5,
		UsageType = "Consumable",
		ExpirationTime = 120,
		ItemCategory = "Traversal",
		DeploymentType = "GridPlacement",
		AssetPath = "Models/Items/ScaffoldingLarge",
	},
	Dynamite = {
		ItemId = "Dynamite",
		DisplayName = "Dynamite",
		Description = "<b>1 Stick of Dynamite</b>\nWhen placed, will create a small explosion after a short time.",
		Cost = 500_000,
		RebirthRequirement = 5,
		PurchaseQuantity = 1,
		MaxStack = 5,
		UsageType = "Consumable",
		ItemCategory = "Explosive",
		ExplosivePower = WallStrengthConfig.EpicWall,
		ExpirationTime = 3,
		DeploymentType = "Throwable",
		AssetPath = "Models/Items/Dynamite",
	},
	DynamiteBundle = {
		ItemId = "DynamiteBundle",
		DisplayName = "Bundle of Dynamite",
		Description = "<b>More Dynamite</b>\nWhen placed, will create a medium explosion after a short time.",
		Cost = 1_000_000,
		RebirthRequirement = 15,
		PurchaseQuantity = 1,
		MaxStack = 3,
		UsageType = "Consumable",
		ItemCategory = "Explosive",
		ExplosivePower = WallStrengthConfig.DivineWall,
		ExpirationTime = 3,
		DeploymentType = "Throwable",
		AssetPath = "Models/Items/DynamiteBundle",
	},
	Bomb = {
		ItemId = "Bomb",
		DisplayName = "Bomb",
		Description = "<b>1 Small Bomb</b>\nWhen placed, will create a large explosion after a short time.",
		Cost = 5_000_000,
		RebirthRequirement = 20,
		PurchaseQuantity = 1,
		MaxStack = 1,
		UsageType = "Consumable",
		ItemCategory = "Explosive",
		ExplosivePower = WallStrengthConfig.CelestialWall,
		ExpirationTime = 3,
		DeploymentType = "Throwable",
		AssetPath = "Models/Items/Bomb",
	},
	MakeshiftStickyBomb = {
		ItemId = "MakeshiftStickyBomb",
		DisplayName = "Makeshift Sticky Bomb",
		Description = "<b>1 Sticky Bomb</b>Adheres to any cave wall surface and explodes after a short time.",
		Cost = 10_000_000,
		RebirthRequirement = 23,
		PurchaseQuantity = 1,
		MaxStack = 2,
		UsageType = "Consumable",
		ItemCategory = "Explosive",
		ExplosivePower = WallStrengthConfig.LegendaryWall,
		ExpirationTime = 3,
		DeploymentType = "SurfacePlacement",
		AssetPath = "Models/Items/MakeshiftStickyBomb",
	},
}

return Items
